loader: do ELF substitutions in memory

lmb opened the pull request:
pkg/bpf: allow overriding constants during load

    Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

loader: remove loader.ELFSubstitutions

    Seems like a needless abstraction, remove it.

    Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

loader: make EndpointHash infallible

    The loader exports a method to turn an endpoint into a hash which is meant
    to change any time the BPF needs to be regenerated. This method is currently
    fallible because of a work around needed because hash.Hash is not easily
    copied. This in turn causes a bunch of awkward error handling where there is
    none needed. It also means that the endpoint code in theory has to deal with 
    the case of datapathRegenCtx.bpfHeaderfileHash being empty.

    Replace the fallible implementation with an infallible one. The new
    implementation doesn't produce the same hashes as the old one, but this
    doesn't matter since the semantics are unchanged. The hashes are used to
    identify templates which aren't reused across agent restarts.

    Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

loader: stop linking template.o into ep.StateDir()

    Compilation of an endpoint program creates a template.o symlink in the
    endpoint state directory. This is problematic since we really should be
    creating the file in the "next" state directory. Overall, the loader should
    not fiddle around in the endpoints state directory.

    Replace the symlink with template.txt in the endpoint state directory, which
    contains the hash of the used template. Importantly, we write this file from
    the endpoint code, not from the loader.

    Move the only user of template.o to template.txt.

    Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

loader: remove ELF substitution logic

    In theory, we need to recompile the BPF for an endpoint any time its
    configuration changes. Compiling is fairly expensive, so a mechanism called
    ELF templating / substitution was developed to avoid recompilation in the
    most common cases.

    1. bpf_lxc.c or bpf_host.c are compiled with a "generic"
      ep_config.h that contains placeholders into
      template.o.

    2. For each endpoint, we replace all the placeholders in
      template.o and write a bpf_lxc.o, etc.

    3. Each time we need to reload the datapath bpf_lxc.o is read
      from disk.

    Replacing the placeholders in step (2) works by parsing the ELF and
    manipulating the string table and data sections directly.

    I believe that this approach was chosen when the loader still relied on
    iproute2 to do all the loading. Now we use cilium/ebpf and the templating
    mechanism is becoming a problem. It is very hard to follow how it works and
    it impedes other refactoring.

    Replace ELF templating with manipulation of ebpf.CollectionSpec directly.

    Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

remove pkg/elf

    There are no users of the package left in the code base. Remove it.

    Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

```release-note
loader: Significantly reduce memory usage during endpoint regeneration
```

lmb committed:
loader: remove loader.ELFSubstitutions

Seems like a needless abstraction, remove it.

Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

lmb committed:
pkg/bpf: allow overriding constants during load

Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

lmb committed:
loader: remove ELF substitution logic

In theory, we need to recompile the BPF for an endpoint any time
its configuration changes. Compiling is fairly expensive, so
a mechanism called ELF templating / substitution was developed to
avoid recompilation in the most common cases.

1. bpf_lxc.c or bpf_host.c are compiled with a "generic"
   ep_config.h that contains placeholders into
   template.o.

2. For each endpoint, we replace all the placeholders in
   template.o and write a bpf_lxc.o, etc.

3. Each time we need to reload the datapath bpf_lxc.o is read
   from disk.

Replacing the placeholders in step (2) works by parsing the ELF
and manipulating the string table and data sections directly.

I believe that this approach was chosen when the loader still
relied on iproute2 to do all the loading. Now we use cilium/ebpf
and the templating mechanism is becoming a problem. It is very hard
to follow how it works and it impedes other refactoring.

Replace ELF templating with manipulation of ebpf.CollectionSpec
directly.

Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

dylandreimerink reviewed (approved):
:+1:

derailed reviewed (approved):
@lmb Nice work and clean up!

lmb committed:
loader: stop linking template.o into ep.StateDir()

Compilation of an endpoint program creates a template.o
symlink in the endpoint state directory. This is problematic
since we really should be creating the file in the "next"
state directory. Overall, the loader should not fiddle
around in the endpoints state directory.

Replace the symlink with template.txt in the endpoint state
directory, which contains the hash of the used template.
Importantly, we write this file from the endpoint code,
not from the loader.

Move the only user of template.o to template.txt.

Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

lmb committed:
remove pkg/elf

There are no users of the package left in the code base. Remove it.

Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

ti-mo commented on pkg/bpf/collection.go:
```diff
@@ -390,11 +413,17 @@ func inlineGlobalData(spec *ebpf.CollectionSpec) error {
+			value := v.value
+			name := strings.TrimPrefix(v.name, constantPrefix)
+			if override, ok := overrides[name]; ok {
+				value = override
```
Can this be done earlier in the function, for example in `for name := range overrides`? I'd merge overrides into `vars` to avoid having this logic here, see my comment below.

lmb replied:
This is nice, thanks!

ti-mo commented on pkg/bpf/collection.go:
```diff
@@ -464,15 +499,16 @@ func globalData(spec *ebpf.CollectionSpec) (varOffsets, error) {
 		// Emit the variable's value by its offset in the datasec.
-		out[vsi.Offset] = value
+		out[vsi.Offset] = globalVar{v.Name, value}
+		names[v.Name] = struct{}{}
```
We could use names as a two-way index:
```suggestion
		names[v.Name] = vsi.Offset
```
Then, in `inlineGlobalData`, we can use this to look up vars by name and merge overrides into values.

ti-mo commented on pkg/bpf/collection_test.go:
```diff
@@ -61,24 +63,24 @@ func TestLoadCollectionResizeLogBuffer(t *testing.T) {
+					// var a
 					0x0, 0x0, 0x0, 0x80,
-					// var 2 has padding since var 3 aligns to 64 bits. Fill the padding
+					// var b has padding since var 3 aligns to 64 bits. Fill the padding
```
```suggestion
					// var b has padding since var c aligns to 64 bits. Fill the padding
```

ti-mo commented on pkg/bpf/collection_test.go:
```diff
@@ -117,9 +120,19 @@ func TestInlineGlobalData(t *testing.T) {
+	}
+
+	delete(spec.Maps, globalDataMap)
+	if err := inlineGlobalData(spec, map[string]uint64{"bogus": 42}); err != nil {
```
Should we leave this nil to isolate the error condition better?
```suggestion
	if err := inlineGlobalData(spec, nil}); err != nil {
```

ti-mo commented on pkg/datapath/loader/netlink.go:
```diff
@@ -74,21 +90,18 @@ func replaceDatapath(ctx context.Context, opts replaceDatapathOptions) (_ func()
 	}
 
+	if opts.elf != "" {
+		return nil, errors.New("opts.elf is set")
```
```suggestion
		return nil, errors.New("opts.elf is not set")
```

lmb replied:
No, the error is correct.

ti-mo replied:
Oh, I see..

ti-mo reviewed (approved):
:rocket: Left a few nits

christarazi reviewed (approved):
ci-structure LGTM
